pkg/skaffold/plugin/shared/server.go

```diff
@@ -0,0 +1,148 @@
+	yaml "gopkg.in/yaml.v2"
+)
+
+// Here is an implementation that talks over RPC
+type BuilderRPC struct {
+	client *rpc.Client
+}
+
+func (b *BuilderRPC) Init(opts *config.SkaffoldOptions, env *latest.ExecutionEnvironment) {
+	// We don't expect a response, so we can just use interface{}
+	var resp interface{}
+	args := InitArgs{
+		Opts: opts,
+		Env:  env,
+	}
+	b.client.Call("Plugin.Init", args, &resp)
+}
+
+func (b *BuilderRPC) Labels() map[string]string {
+	var resp map[string]string
+	err := b.client.Call("Plugin.Labels", new(interface{}), &resp)
+	if err != nil {
+		// You usually want your interfaces to return errors. If they don't,
+		// there isn't much other choice here.
+		logrus.Error("Unable to get labels from server.")
+		panic(err)
+	}
+	return resp
+}
+
+func (b *BuilderRPC) Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]build.Artifact, error) {
+	var resp []build.Artifact
+	if err := convertPropertiesToBytes(artifacts); err != nil {
+		return nil, errors.Wrapf(err, "converting properties to bytes")
+	}
+	args := BuildArgs{
+		ImageTags: tags,
+		Artifacts: artifacts,
+	}
+	err := b.client.Call("Plugin.Build", args, &resp)
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func convertPropertiesToBytes(artifacts []*latest.Artifact) error {
+	for _, a := range artifacts {
+		data, err := yaml.Marshal(a.BuilderPlugin.Properties)
+		if err != nil {
+			return err
+		}
+		a.BuilderPlugin.Contents = data
+		a.BuilderPlugin.Properties = nil
+	}
+	return nil
+}
+
+// Here is the RPC server that BuilderRPC talks to, conforming to
+// the requirements of net/r
+type BuilderRPCServer struct {
+	Impl PluginBuilder
+}
+
+func (s *BuilderRPCServer) Init(args InitArgs, resp *interface{}) error {
+	s.Impl.Init(args.Opts, args.Env)
+	return nil
+}
+
+func (s *BuilderRPCServer) Labels(args interface{}, resp *map[string]string) error {
+	*resp = s.Impl.Labels()
+	return nil
+}
+
+func (s *BuilderRPCServer) Build(b BuildArgs, resp *[]build.Artifact) error {
+	artifacts, err := s.Impl.Build(context.Background(), os.Stdout, b.ImageTags, b.Artifacts)
+	if err != nil {
+		return errors.Wrap(err, "building artifacts")
+	}
+	*resp = artifacts
+	return nil
+}
+
+// InitArgs are args passed via rpc to the builder plugin on Init()
+type InitArgs struct {
+	Opts *config.SkaffoldOptions
+	Env  *latest.ExecutionEnvironment
+}
+
+// BuildArgs are the args passed via rpc to the builder plugin on Build()
+type BuildArgs struct {
+	tag.ImageTags
+	Artifacts []*latest.Artifact
+}
+
+// This is the implementation of plugin.Plugin so we can serve/consume this
+//
+// This has two methods: Server must return an RPC server for this plugin
+// type. We construct a BuilderRPCServer for this.
+//
+// Client must return an implementation of our interface that communicates
+// over an RPC client. We return BuilderRPC for this.
+//
+// Ignore MuxBroker. That is used to create more multiplexed streams on our
```

Comment:
is this something we need in the interface? can it just be removed?

Reply:
I removed the comments bc they're not super necessary, but this is an implementation of the hashicorp plugin interface so I don't think we can remove anything :(